### PR TITLE
Menu api

### PR DESCRIPTION
**Pull request by ggggg**

### CODE REVIEW

File: src/BP-CoreLib/OptionMenu/ActionLabel.cs [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+        public Action<ShPlayer, string> action;
+
+        public ActionLabel(string name, Action<ShPlayer, string> action)
+        {
+            this.name = name;
+            this.action = action;
+        }
+    }
+
+    public class OptionMenu
```
Comment (ggggg) [inline src/BP-CoreLib/OptionMenu/ActionLabel.cs]: Should make it internal tbh

File: src/BP-CoreLib/OptionMenu/ActionLabel.cs [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+using BrokeProtocol.Collections;
+using BrokeProtocol.Entities;
+using BrokeProtocol.Required;
+
+namespace BPCoreLib.OptionMenu
+{
+    public class ActionLabel
+    {
+        public string name;
+        public Action<ShPlayer, string> action;
```
Comment (PLASMAchicken) [inline src/BP-CoreLib/OptionMenu/ActionLabel.cs]: Shouldn't this be Name and Action

File: src/BP-CoreLib/OptionMenu/ActionLabel.cs [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+    public static class Extension
+    {
+        public static void SendOptionMenu(this ShPlayer player, string title, LabelID[] labels, ActionLabel[] actions)
+        {
+            (new OptionMenu(title, labels, actions)).SendToPlayer(player);
+        }
+    }
+
+    public class OnStarted : IScript
```
Comment (PLASMAchicken) [inline src/BP-CoreLib/OptionMenu/ActionLabel.cs]: Why this name?

File: src/BP-CoreLib/BP-CoreLib.csproj [hunk 4]
```diff
@@ -228,6 +228,7 @@
   <ItemGroup>
     <Compile Include="Core.cs" />
     <Compile Include="Interfaces\II18n.cs" />
+    <Compile Include="OptionMenu\ActionLabel.cs" />
```
Comment (PLASMAchicken) [inline src/BP-CoreLib/BP-CoreLib.csproj:232]: Interface?

**Review by PLASMAchicken — CHANGES_REQUESTED** [review]

File: src/BP-CoreLib/BP-CoreLib.csproj (hunk 4 above)
Comment (ggggg) [reply src/BP-CoreLib/BP-CoreLib.csproj:232]: Good idea

**Review by PLASMAchicken — APPROVED** [review]

File: src/BP-CoreLib/OptionMenu/ActionLabel.cs [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+{
+    public class ActionLabel : IActionLabel
+    {
+        public string Name { get; }
+        public Action<ShPlayer, string> Action { get; }
+
+        public ActionLabel(string name, Action<ShPlayer, string> action)
+        {
+            this.Name = name;
+            this.Action = action;
```
Comment (UserR00T) [inline src/BP-CoreLib/OptionMenu/ActionLabel.cs]: `this.` is not required here.
```suggestion
            Name = name;
            Action = action;
```

File: src/BP-CoreLib/OptionMenu/ActionLabel.cs [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+        }
+    }
+
+    internal class OptionMenu
+    {
+        private readonly string _title;
+        public Dictionary<string, ActionLabel> Actions { get; private set; }
+        private readonly LabelID[] _options;
+        private readonly string menuId;
```
Comment (UserR00T) [inline src/BP-CoreLib/OptionMenu/ActionLabel.cs]: Inconsistent field naming. private readonly fields should be camelCased, prefixed with a `_`.
```suggestion
        private readonly string _title;
        public Dictionary<string, ActionLabel> Actions { get; private set; }
        private readonly LabelID[] _options;
        private readonly string _menuId;
```

File: src/BP-CoreLib/OptionMenu/ActionLabel.cs [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+            this.menuId = Guid.NewGuid().ToString();
+            Menus.Add(menuId, this);
+            Actions = new Dictionary<string, ActionLabel>();
+            foreach (var action in actions)
+            {
+                Actions.Add(Guid.NewGuid().ToString(), action);
+            }
+        }
+
+        public static Dictionary<string, OptionMenu> Menus = new Dictionary<string, OptionMenu>();
```
Comment (UserR00T) [inline src/BP-CoreLib/OptionMenu/ActionLabel.cs]: Pull field upwards.

File: src/BP-CoreLib/OptionMenu/ActionLabel.cs [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+
+        internal void SendToPlayer(ShPlayer player)
+        {
+            var actions = new LabelID[this.Actions.Count];
+            int i = 0;
+            foreach (var action in this.Actions)
+            {
+                actions[i++] = new LabelID(action.Value.Name, action.Key);
+            }
```
Comment (UserR00T) [inline src/BP-CoreLib/OptionMenu/ActionLabel.cs]: Could be converted to a for loop so you can inline the `i` variable.

File: src/BP-CoreLib/OptionMenu/ActionLabel.cs [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+            player.svPlayer.SendOptionMenu(_title, player.ID, menuId, _options, actions);
+        }
+    }
+
+    public static class Extension
+    {
+        public static void SendOptionMenu(this ShPlayer player, string title, LabelID[] labels, ActionLabel[] actions)
+        {
+            (new OptionMenu(title, labels, actions)).SendToPlayer(player);
```
Comment (UserR00T) [inline src/BP-CoreLib/OptionMenu/ActionLabel.cs:64]: Instantly disposed POCO's are not a good idea. While I do currently not have a solution in this case, keep this in mind for next time.

**Review by UserR00T — CHANGES_REQUESTED** [review]
Looking good, the-g! I have a few minor changes but nothing too much.

**Review by ggggg — COMMENTED** [review]
Hey, root! Glad to hear from you again. I made some changes to fit your suggestions and also changed the array to take in IEnumerable. I hope this is good now for marge.

**Review by UserR00T — APPROVED** [review]
LGTM! :)

**Pull request merged**